src/wp-includes/media.php

```diff
@@ -5651,104 +5624,154 @@ function wp_get_loading_optimization_attributes( $tag_name, $attr, $context ) {
+		 * Interpret "lazy" as not in viewport. Any other value can be
+		 * interpreted as in viewport (realistically only "eager" or `false`
+		 * to force-omit the attribute are other potential values).
 		 */
-		if ( doing_filter( 'the_content' ) ) {
-			return $loading_attrs;
+		if ( 'lazy' === $attr['loading'] ) {
+			$in_viewport = false;
+		} else {
+			$in_viewport = true;
 		}
+	}
 
-		// Conditionally skip lazy-loading on images before the loop.
-		if (
-			// Only apply for main query but before the loop.
-			$wp_query->before_loop && $wp_query->is_main_query()
+	// Logic to handle a `fetchpriority` attribute that is already provided.
+	if ( isset( $attr['fetchpriority'] ) && 'high' === $attr['fetchpriority'] ) {
+		/*
+		 * If the image was already determined to not be in the viewport (e.g.
+		 * from an already provided `loading` attribute), trigger a warning.
+		 * Otherwise, the value can be interpreted as in viewport, since only
+		 * the most important in-viewport image should have `fetchpriority` set
+		 * to "high".
+		 */
+		if ( false === $in_viewport ) {
+			_doing_it_wrong(
+				__FUNCTION__,
+				__( 'An image should not be lazy-loaded and marked as high priority at the same time.' ),
+				'6.3.0'
+			);
 			/*
-			 * Any image before the loop, but after the header has started should not be lazy-loaded,
-			 * except when the footer has already started which can happen when the current template
-			 * does not include any loop.
+			 * Set `fetchpriority` here for backward-compatibility as we should
+			 * not override what a developer decided, even though it seems
+			 * incorrect.
 			 */
-			&& did_action( 'get_header' ) && ! did_action( 'get_footer' )
-		) {
-			// Increase media count if there are images in header above a certian minimum size threshold.
-			$maybe_increase_content_media_count();
-			return $postprocess( $loading_attrs, true );
+			$loading_attrs['fetchpriority'] = 'high';
+		} else {
+			$in_viewport = true;
+		}
+	}
+
+	if ( null === $in_viewport ) {
+		$header_area = WP_TEMPLATE_PART_AREA_HEADER;
+
+		switch ( $context ) {
+			// Consider elements with these header-specific contexts to be in viewport.
+			case "template_part_{$header_area}":
+			case 'get_header_image_tag':
+				$in_viewport          = true;
+				$maybe_increase_count = true;
+				break;
+			// Count main content elements and detect whether in viewport.
+			case 'the_content':
+			case 'the_post_thumbnail':
+			case 'do_shortcode':
+				// Only elements within the main query loop have special handling.
+				if ( ! is_admin() && in_the_loop() && is_main_query() ) {
+					// Increase the counter since this is a main query content element.
+					$content_media_count = wp_increase_content_media_count();
+
+					// If the count so far is below the threshold, `loading` attribute is omitted.
+					if ( $content_media_count <= wp_omit_loading_attr_threshold() ) {
+						$in_viewport = true;
+					} else {
+						$in_viewport = false;
```

Comment:
By directly increasing the content media count here, rather than doing so below, we're bypassing the size check (and related filter), meaning that smaller images will cause the count to be increased. Is that expected? For consistency, I'd rather keep all the counting in one place.

```suggestion
				// Only elements within the main query loop have special handling.
				if ( ! is_admin() && in_the_loop() && is_main_query() ) {
					// If the count so far is below the threshold, `loading` attribute is omitted.
					if ( $content_media_count < wp_omit_loading_attr_threshold() ) {
						$in_viewport = true;
						$maybe_increase_count = true;
					} else {
						$in_viewport = false;
```

Reply:
We could do that, in fact initially when I wrote the code I was thinking to do that. However, here we also need to know the current count, and the only way to get that count is to call the `wp_increase_content_media_count()` function.

I don't want to get into the weeds of introducing a new function or changing that here, so the only way to not increase the counter here would be to use `$content_media_count = wp_increase_content_media_count( 0 )` here - which works, but is just a bit awkward. But if you're okay with that, I agree that would make things a bit cleaner.

Another nuance though is that we can still not set `$maybe_increase_count = true` here, it would need to be a separate variable like `$increase_count`, because the "maybe" implies that the count should only be increased if the image has a certain minimum size. For content images, that condition is not relevant (something to potentially rethink, but not as part of this ticket, to avoid functional changes), so here we would _always_ want to increase the count. This is just to say that, even if we removed the logic to actually increase the count from here, we would still need to make a small modification to the conditional increase logic at the end of the new code in this function.

Reply:
Ah, yes, that is a bit awkward. I do think we need a way of both getting and setting the state value for the current content media count. I agree that passing `0` is awkward, so perhaps we can come up with a better solution, but I think it would be fine for now if accompanied by an inline comment.

Reply:
Sounds good, updated in https://github.com/WordPress/wordpress-develop/pull/4912/commits/d4365816bbe6ab127802c0e11dca1e06c26543f0.